facilitator/src/manifest.rs

```diff
@@ -216,6 +221,85 @@ impl SpecificManifest {
+                batch_signing_private_key.identifier
+            ))?;
+
+        self.batch_signing_public_keys()?
+            .get(&batch_signing_private_key.identifier)
+            .context(format!(
+                "key identifier {} not present in manifest batch signing public keys",
+                batch_signing_private_key.identifier
+            ))?
+            .verify(&test_message, signature.as_ref())
+            .context(format!(
+                "failed to verify signature over test message with key {}",
+                batch_signing_private_key.identifier
+            ))
+    }
+
+    /// Checks if all of the packet encryption public keys in the manifest
+    /// match one of the provided packet encryption private keys by encrypting a
+    /// random message and decrypting it. Returns an error if any public key
+    /// does not have a corresponding private key.
+    pub fn verify_packet_encryption_keys(
+        &self,
+        packet_encryption_private_keys: &[PrivateKey],
+    ) -> Result<()> {
+        let test_message: Vec<u8> = (0..100).map(|_| rand::random::<u8>()).collect();
+        'outer: for (identifier, csr) in &self.packet_encryption_keys {
+            let public_key = PublicKey::from_base64(&csr.base64_public_key()?).context(format!(
+                "failed to decode packet encryption public key {} from specific manifest",
+                identifier,
+            ))?;
+
+            let encrypted = encrypt_share(&test_message, &public_key).context(format!(
+                "failed to encrypt test message to packet encryption public key {}",
+                identifier,
+            ))?;
+
+            for private_key in packet_encryption_private_keys {
+                match decrypt_share(&encrypted, &private_key) {
+                    Ok(decrypted) => {
+                        // AEAD decryption succeeding but yielding incorrect
+                        // plaintext is incredibly unlikely
+                        if !decrypted.eq(&test_message) {
+                            return Err(anyhow!(
+                                "decrypted text does not match test message for key {}",
+                                identifier
+                            ));
+                        }
+
+                        // AEAD decryption succeeded and the plaintext is good,
+                        // so move on to the next packet encryption public key
+                        continue 'outer;
+                    }
+                    // AEAD decryption failing is expected if we are using the
+                    // wrong private key, so move on to the next one
+                    Err(_) => continue,
+                }
+            }
+
+            // If we made it here, then no private key was able to decrypt the
+            // test message and we have a key mismatch
+            return Err(anyhow!("unable to decrypt test message encrypted with {} with any available packet decryption key", identifier));
```

Comment:
Alternatively, you could bail early if the array is empty. That'd be useful knowledge.

Reply:
I added the length of the decryption key array as you suggested, which I think also catches the case where the array is empty (it'll print 0). In any case, I don't think you'd make it this far if no packet encryption keys were provided because a call to `prio::encrypt::PrivateKey::from_base64` would fail in `facilitator.rs intake_transport_from_args()` if passed the empty string.